### PR TITLE
Changing algorithm of pathset to support more patterns

### PR DESCRIPTION
**Pull request by npamnani**
Matching wildcard rules for exclude path -

%: Match all files and folders for one level.
%%: Match all files and folders recursively.
%abc: Match all within-level ending in "abc".
abc%: Match all within-level starting with "abc".
%abc%: Match all within-level containing sub-string "abc" anywhere.

This patch fixes https://github.com/facebook/osquery/issues/3692 and https://github.com/facebook/osquery/issues/4750

### CODE REVIEW

**Comment by guliashvili** [comment]
ok to test

File: osquery/events/pathset.h [hunk 1]
```diff
@@ -11,114 +11,130 @@
...
+    patterns_.push_back(std::move(path));
   }
 
   bool find(const std::string& str) const {
     auto path = PathType::createPath(str);
 
     ReadLock lock(mset_lock_);
-    if (paths_.find(path) != paths_.end()) {
-      return true;
+    for (auto& pattern : patterns_) {
```
Comment (guliashvili) [inline osquery/events/pathset.h]: const auto&

File: osquery/events/pathset.h [hunk 2]
```diff
@@ -11,114 +11,130 @@
...
- *
- * PathSet can take any of the two policies -
- * 1. patternedPath - Path can contain pattern '%' and '%%'.
- *                    Path components containing only '%' and '%%' are supported
- *                    e.g. '/This/Path/%'.
- *                    Path components containing partial patterns are not
- *                    supported e.g. '/This/Path/xyz%' ('xyz%' will not be
- *                    treated as pattern).
  */
 template <typename PathType>
```
Comment (guliashvili) [inline osquery/events/pathset.h]: Do we need template? Looks like there was no use of the other class which is already removed

File: osquery/events/pathset.h [hunk 3]
```diff
@@ -11,114 +11,130 @@
...
   }
 
   bool find(const std::string& str) const {
     auto path = PathType::createPath(str);
 
     ReadLock lock(mset_lock_);
-    if (paths_.find(path) != paths_.end()) {
-      return true;
+    for (auto& pattern : patterns_) {
+      if (compare(pattern, path)) {
```
Comment (guliashvili) [inline osquery/events/pathset.h]: Maybe matches instead of the compare?

File: osquery/events/pathset.h [hunk 4]
```diff
@@ -11,114 +11,130 @@
...
 
 class patternedPath {
  public:
   typedef boost::tokenizer<boost::char_separator<char>> tokenizer;
   typedef std::vector<std::string> Path;
-  typedef std::vector<Path> VPath;
   struct Compare {
+    bool compareStrings(boost::string_view vlhs,
+                        boost::string_view vrhs) const {
+      if (vlhs[0] == '*' && vlhs[vlhs.size() - 1] == '*') {
```
[the remark this answers is not in the document]
Comment (guliashvili) [reply osquery/events/pathset.h]: vlhs[vlhs.size() - 1] -> vlhs.back()

[the remark this answers is not in the document]
Comment (guliashvili) [reply osquery/events/pathset.h]: even better would be boost::algorithm::starts_with and boost::algorithm::ends_with

File: osquery/events/pathset.h [hunk 5]
```diff
@@ -11,114 +11,130 @@
...
+      } else if (vlhs[vlhs.size() - 1] == '*') {
+        vlhs = vlhs.substr(0, vlhs.size() - 1);
+        if (vlhs.size() <= vrhs.size()) {
+          vrhs = vrhs.substr(0, vlhs.size());
+        } else {
+          return false;
+        }
+      }
+
+      return (vlhs.compare(vrhs) == 0);
```
Comment (guliashvili) [inline osquery/events/pathset.h]: == ?

File: osquery/events/pathset.h [hunk 6]
```diff
@@ -11,114 +11,130 @@
...
+    }
+
     bool operator()(const Path& lhs, const Path& rhs) const {
```
Comment (guliashvili) [inline osquery/events/pathset.h]: Where is it used?

File: osquery/events/pathset.h [hunk 7]
```diff
@@ -11,114 +11,130 @@
...
+#include <boost/utility/string_view.hpp>
 
 #include <osquery/core.h>
 #include <osquery/filesystem.h>
 
 namespace osquery {
 
 /**
- * @brief multiset based implementation for path search.
+ * @brief Vector based implemention for path search.
```
Comment (guliashvili) [inline osquery/events/pathset.h]: Not sure if we need to specific doc. What do you think? ( I mean writing the container type)

File: osquery/events/pathset.h [hunk 8]
```diff
@@ -11,114 +11,130 @@
...
 class patternedPath {
  public:
   typedef boost::tokenizer<boost::char_separator<char>> tokenizer;
   typedef std::vector<std::string> Path;
-  typedef std::vector<Path> VPath;
   struct Compare {
+    bool compareStrings(boost::string_view vlhs,
+                        boost::string_view vrhs) const {
+      if (vlhs[0] == '*' && vlhs[vlhs.size() - 1] == '*') {
+        vlhs = vlhs.substr(1, vlhs.size() - 2);
```
Comment (guliashvili) [inline osquery/events/pathset.h]: Would appreciate size checks, even if we can be sure that size > x from the code flow. .size() - 2 might overflow (its unsigned)

**Comment by guliashvili** [comment]
Thanks for your contribution, it's indeed good feature and will find good use once we make it perfect. @npamnani

File: osquery/events/pathset.h [hunk 9]
```diff
@@ -11,114 +11,130 @@
...
+  Compare compare;
 };
 
 class patternedPath {
  public:
   typedef boost::tokenizer<boost::char_separator<char>> tokenizer;
   typedef std::vector<std::string> Path;
-  typedef std::vector<Path> VPath;
   struct Compare {
+    bool compareStrings(boost::string_view vlhs,
```
[the remark this answers is not in the document]
Comment (npamnani) [reply osquery/events/pathset.h]: It for the performance reason, with string_view it is just pointer manipulation underneath unlike std::string where sub-strings get  generated.

File: osquery/events/pathset.h [hunk 10]
```diff
@@ -11,114 +11,130 @@
...
+                        boost::string_view vrhs) const {
```
[the remark this answers is not in the document]
Comment (npamnani) [reply osquery/events/pathset.h]: done

**Comment by guliashvili** [comment]
@npamnani Do we have the tests, replicating the situation in the issues you pointed at? I think those would be good things to double check.

**Comment by npamnani** [comment]
>  Do we have the tests, replicating the situation in the issues you pointed at? I think those would be good things to double check.

Can you please elaborate a little more?

File: osquery/events/pathset.h (hunk 8 above)
Comment (npamnani) [reply osquery/events/pathset.h]: This condition never going to happen.

File: osquery/events/pathset.h (hunk 7 above)
Comment (npamnani) [reply osquery/events/pathset.h]: I leave this to you. What do you suggest?

File: osquery/events/pathset.h (hunk 6 above)
Comment (npamnani) [reply osquery/events/pathset.h]: 55       if (compare(pattern, path)) {

Comment (npamnani) [reply osquery/events/pathset.h]: functor

File: osquery/events/pathset.h (hunk 5 above)
Comment (npamnani) [reply osquery/events/pathset.h]: done

File: osquery/events/pathset.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (npamnani) [reply osquery/events/pathset.h]: I thought this way clearly shows what my intention is.

File: osquery/events/pathset.h (hunk 2 above)
Comment (npamnani) [reply osquery/events/pathset.h]: We don't need template anymore. I will remove that in my next installment of technical debt repay. :)

File: osquery/events/pathset.h (hunk 1 above)
Comment (npamnani) [reply osquery/events/pathset.h]: done

**Comment by guliashvili** [comment]
@npamnani For example in #4750 very specific corner case has been found.
When one folder(not_to_monitor) matches the pattern(not_to_monitor%) without pattern sign. Looks like it's a good test case for upgraded pathset.

**Comment by npamnani** [comment]
I have put this in the inotify tests file pls check this out -
ec->path = "/home/not_to_monitor/abc";
EXPECT_FALSE(event_pub_->shouldFire(sc, ec));

**Comment by npamnani** [comment]
@guliashvili , I did not realize that I was pushing the empty string. Rectified it. 
Let me know if I missed anything, thanks.

File: osquery/events/pathset.h [hunk 11]
```diff
@@ -11,149 +11,142 @@
...
- *                    Path components containing only '%' and '%%' are supported
- *                    e.g. '/This/Path/%'.
- *                    Path components containing partial patterns are not
- *                    supported e.g. '/This/Path/xyz%' ('xyz%' will not be
- *                    treated as pattern).
  */
 template <typename PathType>
 class PathSet : private boost::noncopyable {
  public:
   void insert(const std::string& str) {
```
Comment (guliashvili) [inline osquery/events/pathset.h]: std::string pattern

File: osquery/events/pathset.h [hunk 12]
```diff
@@ -11,149 +11,142 @@
...
     auto pattern = str;
     replaceGlobWildcards(pattern);
```
Comment (guliashvili) [inline osquery/events/pathset.h]: In case of the 'file_paths_query' and 'file_paths', file_paths parser does this while parsing. Is there any reason for having exception here?
https://github.com/facebook/osquery/blob/master/osquery/config/parsers/file_paths.cpp#L99
https://github.com/facebook/osquery/blob/master/osquery/config/parsers/file_paths.cpp#L129

File: osquery/events/pathset.h [hunk 13]
```diff
@@ -11,149 +11,142 @@
...
   typedef typename PathType::Compare Compare;
-  std::multiset<Path, Compare> paths_;
+  std::vector<Path> patterns_;
   mutable Mutex mset_lock_;
+  Compare compare;
 };
 
 class patternedPath {
  public:
   typedef boost::tokenizer<boost::char_separator<char>> tokenizer;
```
Comment (guliashvili) [inline osquery/events/pathset.h]: It's used only in one place, why not just put the type there now.

File: osquery/events/pathset.h [hunk 14]
```diff
@@ -11,149 +11,142 @@
...
+        }
+      }
+
+      return (pattern == str);
+    }
+
     bool operator()(const Path& lhs, const Path& rhs) const {
-      size_t psize = (lhs.size() < rhs.size()) ? lhs.size() : rhs.size();
+      auto psize = std::min(lhs.size(), rhs.size());
       unsigned ndx;
```
Comment (guliashvili) [inline osquery/events/pathset.h]: size_t

Comment (guliashvili) [reply osquery/events/pathset.h]: ndx is not used outside the loop anymore. lets declare it inside

File: osquery/events/pathset.h [hunk 15]
```diff
@@ -11,149 +11,142 @@
...
+          str = str.substr(0, pattern.size());
+        } else {
+          return false;
+        }
+      }
+
+      return (pattern == str);
+    }
+
     bool operator()(const Path& lhs, const Path& rhs) const {
```
Comment (guliashvili) [inline osquery/events/pathset.h]: Assumption about the specific ordering of the arguments is made. I think this operator concept was useful until there were to classes. Now it looks more confusing.

Same naming will help (pattern, str)

File: osquery/events/pathset.h (hunk 2 above)
Comment (guliashvili) [reply osquery/events/pathset.h]: Of course it's up to you. I appreciate your work

File: osquery/events/pathset.h [hunk 16]
```diff
@@ -11,149 +11,142 @@
...
       for (ndx = 0; ndx < psize; ++ndx) {
-        if (lhs[ndx] == "**" || rhs[ndx] == "**") {
-          return false;
+        if (lhs[ndx] == "**") {
+          return true;
         }
 
-        if (lhs[ndx] == "*" || rhs[ndx] == "*") {
+        if (lhs[ndx] == "*") {
           continue;
```
[the remark this answers is not in the document]
Comment (uptycs-nishant) [reply osquery/events/pathset.h]: This can not happen, Since lhs[ndx] is string not an element of a string. Here string comparison is happening.  e.g.
/xyz/*/uvw 
tokenizer will break this in {"xyz", "*", "uvw"}.

File: osquery/events/pathset.h [hunk 17]
```diff
@@ -11,149 +11,142 @@
...
     boost::char_separator<char> sep{"/"};
     tokenizer tokens(str, sep);
     Path path;
 
     if (str == "/") {
-      path.push_back("");
+      path.push_back("/");
     }
 
     for (std::string component : tokens) {
```
[the remark this answers is not in the document]
Comment (uptycs-nishant) [reply osquery/events/pathset.h]: I see these two statements are same. If Library is a folder then statement should be same. OS will figure out  if Library is folder or not, even if you provide '/' or not. Form exclude perspective 
'/This/Path' or '/This/Path/' and '/This/Path/is' are two different objects. If I say filter out  '/This/Path/' this should not mean filter out '/This/Path/is' as well since for this we have '/This/Path/%'.  
Again there should not be any confusion about path ending with '/' . Since OS will decide if provided object is directory or not, we can not give this a new meaning at application layer.

File: osquery/events/pathset.h (hunk 12 above)
Comment (uptycs-nishant) [reply osquery/events/pathset.h]: This is a utility class. It should take care of this.

File: osquery/events/pathset.h (hunk 13 above)
Comment (uptycs-nishant) [reply osquery/events/pathset.h]: sure

File: osquery/events/pathset.h (hunk 14 above)
Comment (uptycs-nishant) [reply osquery/events/pathset.h]: sure.

File: osquery/events/pathset.h (hunk 11 above)
Comment (uptycs-nishant) [reply osquery/events/pathset.h]: missed out

File: osquery/events/pathset.h (hunk 15 above)
Comment (uptycs-nishant) [reply osquery/events/pathset.h]: sure

File: osquery/events/pathset.h (hunk 12 above)
Comment (guliashvili) [reply osquery/events/pathset.h]: I think, point is that we should keep this classes 'pure' in terms of not to be dependent on the filesystem(replaceGlobWildcards is depended). It makes class easier to test, and also makes more sense in the context I mentioned earlier.

File: osquery/events/pathset.h (hunk 16 above)
[the remark this answers is not in the document]
Comment (uptycs-nishant) [reply osquery/events/pathset.h]: lhs[0] is "a"
lhs[0] != "**" 
lhs[0] != "*"
lhs[0] == rhs[0] == "a"
match (continue)
lhs[1] is "b*c"
lhs[1] != "**" 
lhs[1] != "*"
lhs[1] != rhs[1]  
no match (return)

Result  
lhs = {"a", "b*c", "d"}
rhs = {"a", "k", "d"}
dont match

File: osquery/events/pathset.h [hunk 18]
```diff
@@ -11,149 +11,142 @@
...
  public:
   typedef boost::tokenizer<boost::char_separator<char>> tokenizer;
   typedef std::vector<std::string> Path;
-  typedef std::vector<Path> VPath;
   struct Compare {
+    bool compareStrings(boost::string_view pattern,
+                        boost::string_view str) const {
+      if (pattern[0] == '*' && pattern[pattern.size() - 1] == '*') {
+        pattern = pattern.substr(1, pattern.size() - 2);
+        if (str.size() >= pattern.size() &&
```
Comment (guliashvili) [inline osquery/events/pathset.h]: str.size() >= pattern.size()  not needed

File: osquery/events/pathset.h [hunk 19]
```diff
@@ -11,149 +11,142 @@
...
+            str.find(pattern) != boost::string_view::npos) {
+          return true;
+        }
+        return false;
```
Comment (guliashvili) [inline osquery/events/pathset.h]: pattern = "*a*c*"
str = "abc"

Will return the false for this case. will not it?

File: osquery/events/pathset.h [hunk 20]
```diff
@@ -11,149 +11,142 @@
...
+  Compare compare;
 };
 
 class patternedPath {
  public:
   typedef boost::tokenizer<boost::char_separator<char>> tokenizer;
   typedef std::vector<std::string> Path;
-  typedef std::vector<Path> VPath;
   struct Compare {
+    bool compareStrings(boost::string_view pattern,
```
Comment (guliashvili) [inline osquery/events/pathset.h]: I feel like this function is a very error prone to implement. Can we replace it's functionality using regex?

File: osquery/events/pathset.h (hunk 16 above)
[the remark this answers is not in the document]
Comment (guliashvili) [reply osquery/events/pathset.h]: Oh, I see   your  point now. Thanks

File: osquery/events/pathset.h (hunk 12 above)
Comment (guliashvili) [reply osquery/events/pathset.h]: I understand your point. I think we could benefit more by making this class more testable and assimilating it better into our config parsers.

File: osquery/events/pathset.h (hunk 19 above)
Comment (uptycs-nishant) [reply osquery/events/pathset.h]: Yes, it will return false and it should, since first two characters "ac" != "ab"

File: osquery/events/pathset.h (hunk 18 above)
Comment (uptycs-nishant) [reply osquery/events/pathset.h]: *abcdef*
abcde

in the above example I have to make sure str.size() >= pattern.size() before comparison else there is not point for going for comparison since "abcdef" != "abcde"

**Comment by uptycs-nishant** [comment]
Can you please show me this alog is error prone for one use case using dry run?

**Comment by uptycs-nishant** [comment]
I believe regex will be overkill  here.

**Comment by uptycs-nishant** [comment]
@guliashvili , hey I did not get you last comment "Feedback provided on top".

**Comment by guliashvili** [comment]
@uptycs-nishant 
Sorry for that. Code is too packed with comments right now. Let's do one more round if you decide to improve things, you think need to.

**Comment by uptycs-nishant** [comment]
@guliashvili , please let me know if I have addressed your concerns.

**Comment by guliashvili** [comment]
Awesome work. Thank you. This is fixing a lot of problems and also adds better filtering using better filtering for the folder names.

As you mentioned the issues #3692 and #4750, I think this PR would benefit from having this nice corner cases as the separate tests. Cleaning unused parts of the code in this PR would also be a nice idea. 

compareStrings can also be simplified.
Something like 
int idx = str.find(pattern);
return (idx == 0 || starts_with(pattern, '*') && ...

**Comment by uptycs-nishant** [comment]
> int idx = str.find(pattern);
> return (idx == 0 || starts_with(pattern, '*') && ...

if str = "xyzabc" and pattern = "xyz"
idx = str.find(pattern); 
idx = 0
return (idx == 0); // means pattern matches the string.

Would that be not a wrong result ?

**Comment by guliashvili** [comment]
```
return (idx == 0 || starts_with(pattern, '*') && ...
==
return (idx == 0 || starts_with(pattern, '*') && ( (idx != string::npos && idx + pattern.size() == str.size()) || ends_with(str, '*'))
```
